### PR TITLE
Added 1.14.4 and 1.15 support

### PR DESCRIPTION
**Pull request by NgLoader**
Currently finished

### CODE REVIEW

**Comment by ProgrammerDan** [comment]
Thank you, will review tonight!

**Comment by ProgrammerDan** [comment]
Broad strokes are OK, some details to fine tune -- do you figure you are done with this contribution or do you intend to work on it more? Regardless, thanks for this contribution, will blend with the prior PR, alongside some additional work I intend to begin soon to figure out the longstanding bugs.

File: Plugin/pom.xml [hunk 1]
```diff
@@ -20,7 +20,7 @@
     <dependency>
```
Comment (ProgrammerDan) [inline Plugin/pom.xml:20]: It's likely the POM will need some other changes. The prior PR #235 had the right idea in some places, esp the version increase (although preserve the -SNAPSHOT -- I remove that for releases alone).

File: pom.xml [hunk 2]
```diff
@@ -38,6 +38,7 @@
     <module>v1_12_R1</module>
     <module>v1_13_R1</module>
     <module>v1_13_R2</module>
+    <module>v1_14_4_R1</module>
```
Comment (ProgrammerDan) [inline pom.xml]: Module should be v1_14_R1 -- so far as I can see there are no API changes, just fix releases, so it's just v1_14_R1.

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/BlockInfo.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/**
+ * @author Aleksey Terzi
+ *
+ */
+
+package com.lishid.orebfuscator.nms.v1_14_2_R1;
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/BlockInfo.java]: package and folder should be v1_14_R1, or at the least, consistent with base folder.

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/ChunkCache.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+        return null;
+    }
+
+    private synchronized void closeCacheFilesInternal() {
+        for (RegionFile regionFile : cachedRegionFiles.values()) {
+            try {
+                if (regionFile != null) {
+                	// This lovely piece of work is due to an NMS change in Spigot 1.13.2 without an R increase.
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/ChunkCache.java]: Due to release of obfuscation mapping, I can confidently describe that this extra handler is unnecessary for 1.14, and can be reduce to just the "close" method retrieval and invocation. (line 38987 in the server deobfuscation for 1.14).

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+
+		return true;
+	}
+
+	private static IBlockData getBlockData(World world, int x, int y, int z, boolean loadChunk) {
+		int chunkX = x >> 4;
+		int chunkZ = z >> 4;
+
+		WorldServer worldServer = ((CraftWorld)world).getHandle();
+		// like in ChunkCache, NMS change without R increment.
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java]: will have to investigate bukkit here to see which of the following Methods needs to remain

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+import net.minecraft.server.v1_14_R1.Chunk;
+import net.minecraft.server.v1_14_R1.ChunkProviderServer;
+import net.minecraft.server.v1_14_R1.IBlockData;
+import net.minecraft.server.v1_14_R1.IChatBaseComponent;
+import net.minecraft.server.v1_14_R1.Packet;
+import net.minecraft.server.v1_14_R1.TileEntity;
+import net.minecraft.server.v1_14_R1.WorldServer;
+
+public class NmsManager implements INmsManager {
+	private static final int BITS_PER_BLOCK = 14;
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java]: Not 100% sure but indicators I use, indicate that this bits per block is still 14, so good news there

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+				Material.WHITE_BANNER,
+				Material.WHITE_BED,
+				Material.WHITE_STAINED_GLASS,
+				Material.WHITE_STAINED_GLASS_PANE,
+				Material.WHITE_WALL_BANNER,
+				Material.YELLOW_BANNER,
+				Material.YELLOW_BED,
+				Material.YELLOW_STAINED_GLASS,
+				Material.YELLOW_STAINED_GLASS_PANE,
+				Material.YELLOW_WALL_BANNER
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java]: Some good expansion here

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+	public IChunkCache createChunkCache() {
+		return new ChunkCache(this.maxLoadedCacheFiles);
+	}
+
+	public void updateBlockTileEntity(BlockCoord blockCoord, Player player) {
+		CraftWorld world = (CraftWorld)player.getWorld();
+		// 1.13.2 has made this quite a bit different in later builds.
+		TileEntity tileEntity = null;
+		try {
+			Method getTileEntityAt = world.getClass().getMethod("getTileEntityAt", int.class, int.class, int.class);
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java]: Will have to reference craftbukkit impl to see which one is current for 1.14

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+	public boolean isSign(int combinedBlockId) {
+		return BLOCK_ID_SIGNS.contains(combinedBlockId);
+	}
+
+	public boolean isAir(int combinedBlockId) {
+		return BLOCK_ID_AIRS.contains(combinedBlockId);
+	}
+
+	public boolean isTileEntity(int combinedBlockId) {
+		return Block.getByCombinedId(combinedBlockId).getBlock().isTileEntity();
```
Comment (ProgrammerDan) [inline v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java]: Hopefully the combinedID business still works. Will review.

**Review by ProgrammerDan — CHANGES_REQUESTED** [review]
Some minor changes, some stuff to double check -- thanks for contributing!

**Comment by NgLoader** [comment]
I commit some changes soon 👍

**Comment by NgLoader** [comment]
Currently I'm looking into the ArrayIndexOutOfBoundsException error

**Comment by ProgrammerDan** [comment]
Any updates here, friend?

**Comment by Chevels** [comment]
Any news ? Or OFC is definitively abandonned ? :/

**Comment by ProgrammerDan** [comment]
Nothing new from the contributor of this PR, and as for me I haven't had much time for any of my projects. Not officially abandoned but I would welcome dedicated help.

**Comment by SydMontague** [comment]
I tested the branch locally and it didn't work properly. After some digging I figured out it's because of a protocol change the code doesn't attribute for.
(In detail: light information is not transmitted anymore and a non-air block count field has been introduced)

Based on NgLoader's work I wrote a patch that should make it 1.14.4 compatible, it worked on my local test server but that by no means indicates full compatibility. The patch shouldn't be backward compatible either, so for that some more work is required.

[Orebfuscator114.patch.txt](https://github.com/lishid/Orebfuscator/files/3876734/Orebfuscator114.patch.txt)

**Comment by ProgrammerDan** [comment]
Nice, thank you. Will try to do some testing / integration -- would welcome
a full pr as well.

**Comment by NgLoader** [comment]
I commit some changes soon.
Currently is my pc away for a upgrade.

Thanks @SydMontague

**Comment by NgLoader** [comment]
And @Sydmontague you can commit on my frok so you don't need to send a txt file with changes 😅

**Comment by NgLoader** [comment]
I moved the project to a new location
[Repo](https://github.com/TeamWuffy/Orebfuscator)

**Comment by SydMontague** [comment]
I created a pull request to your repo and also fixed some more issues. The NmsManager did some obsolete reflection that caused the plugin to not work properly in the Nether. In the process I also found a potential NPE I prevented with a simple null check.

The changes I made should now also be backwards compatible...

**Comment by NgLoader** [comment]
Its working for me too, you can test the version [here](https://github.com/TeamWuffy/Orebfuscator/releases/latest)

I'm currently finished
Some small updates coming soon

File: v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java (hunk 7 above)
Comment (NgLoader) [reply v1_14_4_R1/src/main/java/com/lishid/orebfuscator/nms/v1_14_2_R1/NmsManager.java]: currently not, sry

**Comment by NgLoader** [comment]
1.15 update coming soon :smile: currently testing works fine

**Comment by Miky88** [comment]
rip why closed

**Comment by NgLoader** [comment]
The project was moved to Imprex-Development for 1.8-1.15 support.
1.16 ist currently in progress

**Pull request closed without merging**